src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsImpersonationContext.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.ComponentModel;
+using System.Threading;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Security.Principal
+{
+    public class WindowsImpersonationContext : IDisposable
+    {
+        private readonly SafeAccessTokenHandle _previousUserToken = SafeAccessTokenHandle.InvalidHandle;
+
+        private WindowsImpersonationContext() { }
+
+        internal WindowsImpersonationContext(SafeAccessTokenHandle currentUserToken, bool isImpersonating)
+        {
+            if (currentUserToken.IsInvalid)
+            {
+                throw new ArgumentException(SR.Argument_InvalidImpersonationToken);
+            }
+
+            if (isImpersonating)
+            {
+                // if we're in a impersonated context we need to duplicate current token to allow revert to correct identity 
+                // after impersonation
+                _previousUserToken = WindowsIdentity.DuplicateAccessToken(currentUserToken);
+            }
+        }
+
+        public void Undo()
+        {
+            if (!Interop.Advapi32.RevertToSelf())
+            {
+                Environment.FailFast(new Win32Exception().Message);
+            }
+
+            WindowsIdentity.s_currentImpersonatedToken.Value = null;
+
+            // revert impersonating token and impersonate previous identity
+            if (!_previousUserToken.IsInvalid)
+            {
+                if (!Interop.Advapi32.ImpersonateLoggedOnUser(_previousUserToken))
+                {
+                    throw new SecurityException(SR.Argument_ImpersonateUser);
+                }
+
+                // reset AsyncLocal to allow impersonation context flow in case of async/await
+                WindowsIdentity.s_currentImpersonatedToken.Value = _previousUserToken;
+            }
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (_previousUserToken != null && !_previousUserToken.IsClosed)
+                {
+                    Undo();
+                }
```

Comment:
Seems to be missing the `_previousUserToken.Dispose()` that was present in netfx.